### PR TITLE
build/manifest: print whole manifest variable as it is

### LINKED ISSUE #72: Update manifest.go to include entire plugin manifest.
Update `build/manifest.main.go` to dump the entire manifest from `plugin.json` into `server/manifest.go` instead of just the two fields listed there now. The declared type can just be `model.Manifest` from the `github.com/mattermost/mattermost-server/model` package.

This will enable plugins to reference their own manifests programmatically without having to parse the `plugin.json` manually.

----
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA: https://mattermost.atlassian.net/browse/MM-20035

### PR DESCRIPTION
**Pull request by ilgooz**
#### Summary

Print whole manifest variable as it is while generating manifest files for both JS and Go sources.

* added new `vargen` package to do code generation for Go's runtime variables.

* notice that the code generation package _(github.com/dave/jennifer/jen)_ used in `vargen` package has two cons:
  * it reorders struct types' fields by alphabet instead of leaving it as it appears during struct's initialization.
  * it does not put new lines for evert struct field if struct is initialized with under 2 fields. this might be annoying for some.

* we could fix cons described above by creating our own code generator that directly builds code by bare writing to a buffer or by using `go/ast` but both ways will add more code to maintain. and this might not make sense for some small advantage.

* notice that generated Go variable may contain slice, array, map and struct values with no members. this is valid because they must have been initialized with no members. zero values on the other hand are discarded and won't appear in the generated code.

* notice that generated JS variable may contain zero values. this is valid because code generation for JS made by using JSON marshaling and `encoding/json` removes empty values only when `omitempty` tag is used.

* this commit introduces breaking changes since generated `manifest.js` and `manifest.go` files are now exporting `model.Manifest` type.

* ran go mod tidy.

#### Related Ticket
fixes #72.

### CODE REVIEW

**Comment by ilgooz** [comment]
It seems CI cannot make tests pass where it's all OK at my local machine. I'll check and try to fix CI errors.

**Comment by lieut-data** [comment]
Leaving the full review to @iomodo and @mickmister, but I wonder if there's another, third option that might be simpler: use [ManifestFromJson](https://github.com/mattermost/mattermost-server/blob/master/model/manifest.go#L215-L219) and just write out the raw JSON as an inline parameter to same? One downside is the need for the plugin to have to "parse" the JSON on each start, but then we wouldn't need any extra tools to process it.

**Comment by ilgooz** [comment]
> Leaving the full review to @iomodo and @mickmister, but I wonder if there's another, third option that might be simpler: use [ManifestFromJson](https://github.com/mattermost/mattermost-server/blob/master/model/manifest.go#L215-L219) and just write out the raw JSON as an inline parameter to same? One downside is the need for the plugin to have to "parse" the JSON on each start, but then we wouldn't need any extra tools to process it.

Totally right. I might be overcomplicated things with `vargen`. The only pros it provides is, it just renders a usual Go code which looks better for the eyes.

On the other hand a solution by parsing JSON encoded manifest to _model.Manifest_ var is the easiest and less error prone for development and maintenance work. We can do the decoding within `init()` func and it'll get cached to a _var_ for the program's runtime.

It is also possible to print this JSON data in a human friendly way so developers still can look at the _manifest.go_ to get an overall idea about how it looks like.

**Comment by lieut-data** [comment]
> It is also possible to print this JSON data in a human friendly way so developers still can look at the manifest.go to get an overall idea about how it looks like.

Agreed, let's definitely render this in a human readable way. I do like the simplicity of the rendered Go code provided by `vargen`, but it does seem to require a lot of extra overhead. I had somewhat expected it (or some similar package) to be able to write out an arbitrary struct without further configuration: that would be the ideal solution.

**Comment by ilgooz** [comment]
@lieut-data I'll do the updates to use _JSON_ approach over _vargen_ for the _Go_ version once we have a full agreement!

Hey @prapti @iomodo @mickmister what are your ideas about this?

**Comment by ilgooz** [comment]
> I had somewhat expected it (or some similar package) to be able to write out an arbitrary struct without further configuration: that would be the ideal solution.

Before coming up with _vargen_, I researched for a tool to do this for us, to write Go structs from runtime values and [pretty](https://github.com/kr/pretty) is the closet one to what we would look for. 

But even it has a few cons, it does not produce a valid Go code for empty _maps_ and it puts some extra type indicator for values that we usually don't have in a regular source Go code. I didn't prefer to create a PR to fix these because the project seems inactive now.

**Comment by iomodo** [comment]
Great PR @ilgooz! Go code does look nice, but I'd still vote for a JSON version - a much, much simpler code.

**Comment by ilgooz** [comment]
ready to review.

switched to JSON unmarshaling by 8db2d85098ad877f4a253cd97b91f2d67834d798.

File: build/manifest/main.go [hunk 1]
```diff
@@ -102,20 +109,33 @@ func dumpPluginVersion(manifest *model.Manifest) {
 
 // applyManifest propagates the plugin_id into the server and webapp folders, as necessary
 func applyManifest(manifest *model.Manifest) error {
+	// generate JSON representation of Manifest.
+	// * JSON is very similar and compatible with JS's object literals. so, what we do here
+	//   is actually JS code generation.
+	// * this JSON ouput will also be used by the Go version to parse it internally to
+	//   expose a *model.Manifest.
+	manifestBytes, err := json.MarshalIndent(manifest, "", " ")
```
Comment (lieut-data) [inline build/manifest/main.go]: Could we switch this to indent with two spaces, and match other `.json` files as used in our various repositories?
```suggestion
	manifestBytes, err := json.MarshalIndent(manifest, "", "  ")
```

File: build/manifest/main.go [hunk 2]
```diff
@@ -102,20 +109,33 @@ func dumpPluginVersion(manifest *model.Manifest) {
 
 // applyManifest propagates the plugin_id into the server and webapp folders, as necessary
 func applyManifest(manifest *model.Manifest) error {
+	// generate JSON representation of Manifest.
+	// * JSON is very similar and compatible with JS's object literals. so, what we do here
+	//   is actually JS code generation.
+	// * this JSON ouput will also be used by the Go version to parse it internally to
```
Comment (mickmister) [inline build/manifest/main.go]: Very small nit:

```suggestion
	// * this JSON output will also be used by the Go version to parse it internally to
```

**Comment by ilgooz** [comment]
> Not blocking on the PR, but I'm wondering why we parse the manifest on compile time. We already include it into the bundle and could parse the JSON on run time. This would remove the need of a auto generated `server/manifest.go` which would resolve the pain point that `server/manifest.go` changes on every release. At least @levb would love this 😉

I also find this a better and flexible approach but I think we should be providing these kind of functionalities with language SDKs.

For ex:

1- go: `plugin.Manifest() *model.Manifest` > makes a lookup to find `manifest.json` in plugin's root,  parses, caches and returns manifest.

2- js: `plugin.Manifest(): Manifest` > same as the Go version. I don't know if we have but we need a public MM npm package (js SDK) to enable this.

**Comment by lieut-data** [comment]
@hanzei / @ilgooz, on the subject of "why", this idea originated from https://github.com/mattermost/mattermost-plugin-demo/pull/74#discussion_r344851228. One cannot resolve one's own bundle without first making an RPC call (or speculating where on disk you might find it). This seems counterintuitive, hence just making it explicit.

@ilgooz, thanks for updating the spacing! I don't see the actual generated files having been updated yet, though, and I note that our `linter` seems to be complaining about wanting 4 spaces, plus wanting single quotes, etc. Perhaps we converge on 4, regenerate, and split apart the Go-generated code from the JS-generated code to achieve linting satisfaction?

**Comment by levb** [comment]
> ... We already include it into the bundle and could parse the JSON on run time. This would remove the need of a auto generated `server/manifest.go` which would resolve the pain point that `server/manifest.go` changes on every release. At least @levb would love this 😉

(Sorry I am not in the context of this PR, so ignore if irrelevant)
I always thought that the manifest files were a redundant hack - would love to see them disappear. JS could include plugin.json natively, and in golang it can be pre-set in p.MattermostPlugin by the time the plugin is loaded.

**Comment by ilgooz** [comment]
> @ilgooz, thanks for updating the spacing! I don't see the actual generated files having been updated yet, though, and I note that our linter seems to be complaining about wanting 4 spaces, plus wanting single quotes, etc. Perhaps we converge on 4, regenerate, and split apart the Go-generated code from the JS-generated code to achieve linting satisfaction?

I was holding my update to see if I can come up with an elegant solution but seems like running multiple regexes are the only option that I can find for now.

Pushing the changes with 147412b to not block this PR.

**Comment by lieut-data** [comment]
@levb, agree that we might find a way to supply the manifest entirely at runtime. For now, this sticks with the current approach of building into the binary, but supplies the whole manifest instead of just a few fields, avoiding the need to lookup the manifest in related API calls.

File: go.sum [hunk 3]
```diff
@@ -111,6 +111,7 @@ github.com/gomodule/redigo v2.0.0+incompatible/go.mod h1:B4C85qUVwatsJoIUNIfCRsp
 github.com/google/btree v0.0.0-20180813153112-4030bb1f1f0c/go.mod h1:lNA+9X1NB3Zf8V7Ke586lFgjr2dZNuvo3lPJSGZ5JPQ=
 github.com/google/btree v1.0.0/go.mod h1:lNA+9X1NB3Zf8V7Ke586lFgjr2dZNuvo3lPJSGZ5JPQ=
 github.com/google/go-cmp v0.2.0/go.mod h1:oXzfMopK8JAjlY9xF4vHSVASa0yLyX7SntLO5aqRK0M=
+github.com/google/go-cmp v0.3.0 h1:crn/baboCvb5fXaQ0IJ1SGTsTVrWpDsCWC8EGETZijY=
```
Comment (hanzei) [inline go.sum]: Why are there changes there?

File: webapp/src/manifest.js [hunk 4]
```diff
@@ -1,4 +1,28 @@
 // This file is automatically generated. Do not modify it manually.
 
-export const id = 'com.mattermost.plugin-starter-template';
-export const version = '0.1.0';
+const manifest = JSON.parse(`
```
Comment (hanzei) [inline webapp/src/manifest.js]:
```suggestion
export const manifest = JSON.parse(`
```

is one way to fix this

Comment (lieut-data) [reply webapp/src/manifest.js]: The manifest was actually exported as the default, but the test wasn't updated. I've actually updated to export both the manifest and the old id and version just to make migration easier for anyone updating their template.

File: go.sum (hunk 3 above)
Comment (lieut-data) [reply go.sum]: Reverted, then build and ran `go mod tidy` -- changes pushed.

**Comment by jasonblais** [comment]
@prapti Quick reminder to help with this PR testing when you have time :)

**Pull request merged**